### PR TITLE
Track user modifications

### PR DESCRIPTION
**Pull request by mkangia**
https://dimagi-dev.atlassian.net/browse/ICDS-1505

##### SUMMARY
There is a need to audit changes done to a user.
[Creation](https://github.com/dimagi/commcare-hq/pull/27755) and [Deletion](https://github.com/dimagi/commcare-hq/pull/27835)(soft/permanent) has already been covered.

Tracking updates can get challenging and heavy, so this PR just sets up the structure to track updates to user models for the attributes we want to track.

The ticket currently specifies role assignments to track so this PR focusses on domain membership(s). 

- [x] find a way to keep original doc version
- [x] add log entries on save
- [ ] pass user making the change in case of async tasks
- [ ] update/skip existing tests to avoid overhead
- [ ] add test

The log entries will look something like
Screenshot 2020-07-07 at 6 59 18 PM

### CODE REVIEW

**Review by esoergel — COMMENTED** [review]
This approach adds automatic sniffers on the user model, which provides:

* a near-guarantee that every user save will be accounted for
* code that modifies users generally doesn't need to worry about this

That's certainly attractive, but it comes at a cost:

* complex code to figure out what happened retroactively
* additional overhead in dealing with audited objects (it doubles the memory footprint, plus the overhead of `deepcopy` and of the comparision code)
* this stuff can't be skipped even in workflows where it's not necessary (or at least it's hard to do so)

---------------------

An alternative would be to push the auditing logic into the places where users are updated, so when a user is edited in the UI, that form or import or whatever is responsible for recording the edit.  The benefits of that are:

* The model stays simpler
* The auditing code can be pretty straightforward
* Model access that doesn't need to save won't face this overhead

The major drawback of this approach is that it'd be easy to miss a place where the model is updated, or forget to log this in future code.

----------------------------

You can certainly make a case for either approach, but I think my preference is for the latter.  Particularly for users, given how they are used.  Users are frequently accessed, and less frequently modified.  The user models also store way more than they should, and many of the updates aren't relevant.  For instance, the `devices` and `reporting_metadata` properties are frequently modified, but they're internal properties, and I don't think those modifications need to be recorded.  What we need to record is users making modifications to a user model, so it makes sense to have that code do `user.save(); log_user_change()`.

We also do a bunch of stuff on save already - I recall looking into it a couple years ago and finding that a single user save took like 2 seconds.  I see there are 5 signals that are called synchronously whenever a user is saved.  This kind of bloat tightly couples together parts of the codebase, making it difficult to work on discrete components in isolation.

It's certainly a hard problem - IMO it'd be a good candidate for a CEP so others have the opportunity to weigh in.

**Review by snopoke — COMMENTED** [review]
do we intend tracking every single field? Considering that we have a lot of data in user models including analytics data I don't think it's wise to take this strategy. Either we should whitelist the fields that we want to monitor or else we should more manually in the views as is done here: https://github.com/dimagi/commcare-hq/blob/b634a10eed7684ca05908ad24336a3b5086bb3fe/corehq/apps/hqadmin/views/users.py#L146

**Comment by mkangia** [comment]
> Either we should whitelist the fields that we want to monitor 

@snopoke we will only track what we need as set [on the model](https://github.com/dimagi/commcare-hq/pull/28058/files#diff-16ce9f2eb3be0695a91eca75714bb3b7R1087)

**Comment by mkangia** [comment]
thanks for the inputs @esoergel 

Those pros/cons are pretty valid. Just following up on some
> a near-guarantee that every user save will be accounted for

That is actually one big factor. I have been careful about not letting this slip out anywhere in code. We dont want to end up in a situation where we don't have an audit of an edit because the code changes didn't cover it.

> additional overhead in dealing with audited objects (it doubles the memory footprint, plus the overhead of deepcopy and of the comparision code)

we are keeping a copy of the json doc of the object. Is that big enough of a memory change to worry about?

> this stuff can't be skipped even in workflows where it's not necessary (or at least it's hard to do so)

nothing except the copy of object is done unless the user is actually saved. It should be possible to allow a mechanism to skip doing that as well, which would need the code loading the user to be careful they pass in the necessary argument/setup, which might not go well.

Overall, I was thinking of such audit tools that sit in the background and handle the tracking for us with minimal changes needed in the model or other code working with the objects.

File: corehq/ex-submodules/auditcare/utils/models.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        for attr_to_check in self.model_object.track_updates_for:
+            new_value = self.model_object._doc.get(attr_to_check)
+            old_value = self.model_object._original.get(attr_to_check)
+            if old_value and not new_value:
+                changes[attr_to_check]['delete'] = old_value
+            elif new_value and not old_value:
+                changes[attr_to_check]['create'] = new_value
+            else:
+                if hasattr(self.model_object, f"compare_{attr_to_check}"):
+                    attr_changes = getattr(self.model_object, f"compare_{attr_to_check}")(self.model_object._original)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/ex-submodules/auditcare/utils/models.py:47]: Thanks for sharing that Ethan. I agree we should avoid it till we can, I have found myself lost at times where this is done in HQ itself and reading this post makes me think about all strength of ruby programming, being questioned. 
I can't think of a better way here to let it be flexible and allow model to choose what they want to track, however they do suggest adding "comments" where dynamic programming is a good approach. Foutunately, the use is going to be limited to a single class calling it, so should be easy to track.

**Comment by snopoke** [comment]
I'm very much with @esoergel that doing it manually in the places where user actually changes the model is a better approach. There are only a few places where a user can actually be edited so I don't think it's hard to make sure those are all covered. 

As it stands this code will perform the deepcopy on every single user lookup, including cache lookups (and I think will also store the copied data in the cache). Since we do a user lookup on every single request that is going to account for a significant change both in terms of processing and also in cache usage.

**Comment by mkangia** [comment]
> As it stands this code will perform the deepcopy on every single user lookup, including cache lookups (and I think will also store the copied data in the cache)

@snopoke currently the copy is added after fetching the object from cache, that was precisely the reason to avoid saving it like that in cache. So cache won't get overloaded.

Is doing a deepcopy of a user doc that heavy? I fail to understand how that is going to add to load, even if its happening on each user fetch. Can one help understant that bit please?

**Comment by snopoke** [comment]
| @snopoke currently the copy is added after fetching the object from cache, that was precisely the reason to avoid saving it like that in cache. So cache won't get overloaded.

You're doing a deepcopy inside the `wrap` function which get's called before caching.

| Is doing a deepcopy of a user doc that heavy? I fail to understand how that is going to add to load, even if its happening on each user fetch. Can one help understant that bit please?

deepcopy on a user doc takes about 45ms. At 250r/s that adds ~1s of processing time per second of requests across the webworker nodes. Whether or not you feel that's a significant amount or not it's still work that is irrelevant for 99.9% of requests.

**Comment by mkangia** [comment]
Thanks for the numbers to explain that, @snopoke. That is surely much heavier than i thought it would be.

> You're doing a deepcopy inside the wrap function which get's called before caching.

I guess I am checking this incorrectly, but even earlier i could not find `_original` on the cached returns making me think that it was no stored in the cached object. I guess it's there in the cache but is not put back on the object since its not a property.

```
In [15]: ccu = CommCareUser.get_by_username("test@test.commcarehq.org")

In [16]: ccu._original is None
Out[16]: False

In [17]: ccu_from_cache = cache.get(CommCareUser.get_by_username.get_cache_key(CommCareUser, username="test@test.commcarehq.o
    ...: rg"))

In [18]: ccu_from_cache._original
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-18-07e0880f8854> in <module>()
----> 1 ccu_from_cache._original

/Users/manishkangia/Projects/dimagi/commcare-hq/corehq/apps/users/models.py in __getattr__(self, item)
   1667
   1668         raise AttributeError("'{}' object has no attribute '{}'".format(
-> 1669             self.__class__.__name__, item))
   1670
   1671     def _get_perm_check_fn(self, perm):

AttributeError: 'CommCareUser' object has no attribute '_original'
```

File: corehq/ex-submodules/auditcare/utils/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/ex-submodules/auditcare/utils/models.py:47]: An alternative would be to store a tuple in `track_updates_for` - something like:
```python
for attr, comparison_fn in self.model_object.track_updates_for:
```
Depending on context, that could be just a string that matches the method name, or more ideally an actual reference to it.  Even if this ends up looking repetitive, it'd be easier to trace usage of those methods.

**Comment by mkangia** [comment]
Closing in favor of https://github.com/dimagi/commcare-hq/pull/28631

**Pull request closed without merging**